### PR TITLE
feat(auth): Backfill accountAuthorizations with refresh tokens

### PR DESCRIPTION
**Pull request by nshirley**
Because:
 - We want to be able to backfill refreshTokens into the accountAuthorizations table

This commit:
 - Adds a backfill script to walk the refreshTokens table, inserting a new accountAuthoriztions for the most recent token/scope/service combo
 - Adds unit tests for the backfill script
 - Includes the seed-test-data used for integration testing of the backfill
 - Includes the test cases in MD format for review

Closes: FXA-12932

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).
- [x] I have manually reviewed all AI generated code.

## How to review (Optional)

I've included a few scripts with this to help in testing (all of which can be removed or kept before merge). First, there is a `backfill-account-authorizations-test-cases.md` to attempt capturing all possible cases that we need the backfill to account for and test. These are used in `seed-test-data.ts` to generate mock data in the local refreshTokens table. From there, we can run the backfill, and then test the results with the SQL outlined in the test-cases doc.

So:
1. Start infra
2. CD to `packages/fxa-auth-server/`
3. run the seed-data `node -r esbuild-register scripts/backfill-account-authorizations/seed-test-data.ts`
    - Optionally, `--clean` to remove data if you need
    - Or, `--volume-count <num>` to seed large amounts of data to test batching in backfill
4. run the backfill `node -r esbuild-register scripts/backfill-account-authorizations/backfill-account-authorizations.ts`
    - Use `--dry-run` to see what it would insert without writing records (still queries refreshTokens, 
    - Or, target a single service `--service relay`. Of note, this WILL still scan the full refreshTokens table, but only cares about the noted service. This _could_ be offloaded to SQL, but offers little in performance gains doing so
5. Check the results using the SQL in `backfill-account-authorizations-test-cases.md`!

The output from the backfill will show several important data points, including run time, number of skipped tokens, and counts broken down by service - this is the result of locally backfilling ~1M refreshTokens. So, assuming linear scaling, at ~70M prod refreshTokens, we can estimate about 140 minutes to scan and process the full table
image

And, for prosperity, here's the output when targeting a single service
```
{"Timestamp":1778172264072000000,"Logger":"backfill-account-authorizations","Type":"backfill.complete","Severity":6,"Pid":611,"EnvVersion":"2.0","Fields":{"tokensScanned":100027,"rowsUpserted":20018,"skippedAmbiguous":0,"skippedNoMatch":70007,"totalDurationMs":11151,"totalDurationHuman":"11s","countsByService":"{\"relay\":20018,\"smartwindow\":0,\"sync\":0,\"vpn\":0}"}}
```

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

### CODE REVIEW

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md
Comment (nshirley) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md:1]: I'll remove this, it's just a way to keep track of the various test cases from seed data locally. Might be worth keeping though 🤷

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+}
+
+/**
+ * Splits the full BINARY(32) token keyspace into `workerCount` equal ranges
+ * and returns the [startCursor, endCursor] pair for `workerIndex`.
+ *
+ * Tokens are SHA256 hashes so they distribute uniformly, making byte-boundary
+ * splits an effective way to parallelize without hotspots.
+ */
+function workerCursors(
```
Comment (nshirley) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:267]: We might not need the multiple worker functionality with this. It's nice to have, but might require more setup within the k8 cluster - whoever reviews this, let me know your thoughts!

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+    `INSERT INTO accountAuthorizations (uid, scope, service, authorizedAt, lastUsedAt)
+     VALUES ${placeholders}
+     ON DUPLICATE KEY UPDATE
+       lastUsedAt   = GREATEST(lastUsedAt,   VALUES(lastUsedAt)),
+       authorizedAt = LEAST(authorizedAt, VALUES(authorizedAt))`,
+    values
+  );
+}
+
+function writeCheckpoint(
```
Comment (nshirley) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: This is a nice to have, but if the container is shutdown then the file is lost. A better option would be storing it in mysql, but it's not super necessary.

Luckily, the inserts are idempotent, so there's no risk in start at the beginning again if the script crashes. It just means we have to scan the table again

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+  const clientHex = clientId.toString('hex').toLowerCase();
+  const scopes = new Set(scopeStr.split(/\s+/).filter(Boolean));
+
+  return services.filter(
+    (svc) =>
+      svc.clientIdSet.has(clientHex) || scopes.has(svc.authorizationScope)
+  );
+}
+
+async function batchUpsert(
```
Comment (nshirley) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: I can probably use the work from https://github.com/mozilla/fxa/pull/20500 to use the functions there for upsert. No sense in reinventing the wheel, this was just place holder while working on the backfill

Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: So, circling back to this, it's probably not necessary to use the underlying db/mysql functions added as part of the work to create the table.

Few reasons for this:
- The db layer is responsible for app behavior, which has an interesting caveat that it uses an `INSERT IGNORE` such that the first authz grant becomes the 'oldest'. This is correct as far as the table / business logic goes, but doesn't help us with the backfill. When backfilling, we may find a token for a user who granted access in 2025, _then_ later in processing we find a token granted (same scopes, etc.) for 2024. We want the oldest to be inserted, hence why we use `ON DUPLICATE KEY UPDATE`
- The other problem: the db layer does a single insert, and we're wanting to backfill millions of records so we do bulk updates with `INSERT ... VALUES (), ()`. If we switch to using the single insert we significantly increase run time
- Lastly, we _could_ create a function in mysql db layer like `_bulkUpsertAccountAuthorizationsForBackfill` but this then adds a backfill only function next to 'production' code and we probably don't want that. (though, maybe there's an argument for making this for RP's to be able to quickly de-authorize multiple users?)

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+| T21 | Idempotency                | Script run twice on identical data                                                                                            | Result identical to single run                                                                                                                                 |
+| T22 | Strict matching            | 1 token, relay scope, clientId that is **not** a relay minter                                                                 | 0 rows — scope alone is not proof of authorization (matches `recordAuthorizationOnLogin`)                                                                      |
+| T23 | Strict matching            | 1 token, no service scope, clientId that maps to **multiple** services (e.g. mobile Firefox in relay+vpn+sync)                | 0 rows — multi-service clientId without scope is ambiguous, no fallback                                                                                        |
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md:39]: Maybe "no service scope" is throwing me off, but if there's a refresh token associated with a client ID with multiple scopes, we want to populate the table. For example, [this data returned from connected services](https://github.com/mozilla/fxa/blob/main/packages/fxa-settings/src/components/Settings/ConnectedServices/mocks.ts#L351) is associated with the Firefox iOS client ID but it has the Sync scope as well as the Relay scope. That should be considered authorization for sync and relay.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+- **T22–T25 (strict matching edge cases)** are documented but not yet in the seed. To verify, add tokens with:
+  - T22: relay scope + an unrelated clientId (uid `aa…0022`)
+  - T23: profile scope + a multi-service clientId like mobile Firefox `1b1a3e44c54fbb58` (uid `aa…0023`)
+  - T24: profile scope + Firefox Desktop `5882386c6d801776` (uid `aa…0024`)
+  - T25: relay scope + Firefox Desktop `5882386c6d801776` (uid `aa…0025`)
+- **T26–T34** are run-mode tests (dry run, resumability, volume, service filter) verified manually via flags and log inspection.
+
+## Open Questions
+
+- **Sync (`allowSilentExchange: false`)**: Should Sync refresh tokens be backfilled? The flag governs future silent exchanges, not historical state. If excluded, add `backfill: false` to the service config. Needs confirmation before implementation.
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md]: This should be: should Sync ***desktop*** refresh tokens be backfilled?

[related note here under caveats](https://github.com/mozilla/fxa/pull/20500#pullrequestreview-4238172630)

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+  // single-connection pattern doesn't auto-reconnect; the next query reject
+  // will propagate up and exit non-zero — resume from the last checkpoint.
+  conn.on('error', (err) => {
+    statsd?.increment('account_authz.backfill.connection_error', 1, {
+      code: (err as { code?: string }).code ?? 'unknown',
+    });
+    log.error('backfill.connection_error', {
+      code: (err as { code?: string }).code,
+      err: errMessage(err),
+      msg: 'Connection lost mid-run; resume with --start-cursor from the last checkpoint.',
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: Smol nit:  maybe on function init display something about this, in case for some reason the error message noting it doesn't get written to the console?

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 7]
```diff
@@ -0,0 +1,695 @@
...
+    const workerIndex = parseInt(program.workerIndex ?? '0', 10);
+    let cursors: { startCursor: Buffer; endCursor: Buffer };
+    try {
+      cursors = workerCursors(workerIndex, workerCount);
+    } catch (err) {
+      log.error('backfill.worker_config_error', { err: errMessage(err) });
+      return 1;
+    }
+    startCursor = cursors.startCursor;
+    endCursor = cursors.endCursor;
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:667]: Does this mean these get reset to defaults if the `--worker-count` option is passed?

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 8]
```diff
@@ -0,0 +1,695 @@
...
+     ON DUPLICATE KEY UPDATE
+       authorizedAt = LEAST(authorizedAt, VALUES(authorizedAt))`,
+    values
+  );
+}
+
+function writeCheckpoint(cursor: Buffer, filePath: string, log: Logger): void {
+  try {
+    fs.mkdirSync(path.dirname(filePath), { recursive: true });
+    fs.writeFileSync(filePath, cursor.toString('hex'), 'utf8');
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:241]: Does this need to be synchronous?

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 9]
```diff
@@ -0,0 +1,695 @@
...
+    r.authorizedAt,
+  ]);
+
+  await query(
+    `INSERT INTO accountAuthorizations (uid, scope, service, authorizedAt)
+     VALUES ${placeholders}
+     ON DUPLICATE KEY UPDATE
+       authorizedAt = LEAST(authorizedAt, VALUES(authorizedAt))`,
+    values
+  );
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:235]: Claude noted: "No retry on transient MySQL errors. A single packet hiccup kills the whole run; you resume from cursor, so it's annoying-not-broken. A small retry-with-backoff around fetchBatch/batchUpsert would let it ride through blips."

Since we have a lot of refresh tokens in our DB the chances seem high of this happening for at least some of these but also it'll get taken care of with a rerun, so 🤷‍♀️

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 10]
```diff
@@ -0,0 +1,695 @@
...
+ * Tokens are SHA256 hashes so they distribute uniformly, making byte-boundary
+ * splits an effective way to parallelize without hotspots.
+ */
+function workerCursors(
+  workerIndex: number,
+  workerCount: number
+): { startCursor: Buffer; endCursor: Buffer } {
+  // Bounded by the size of the first byte of the keyspace (256 values). More
+  // than 256 workers would produce zero-byte ranges that overlap or skip rows.
+  if (!Number.isInteger(workerCount) || workerCount < 1 || workerCount > 256) {
```
Comment (LZoog) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:273]: How many workers do we think we'll want to use in prod?

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+ *   # Parallel workers — splits the token keyspace evenly across N processes.
+ *   # token is a SHA256 hash so values distribute uniformly across the key space.
+ *   # Run each command in a separate terminal / tmux pane:
+ *   node -r esbuild-register scripts/backfill-account-authorizations/index.ts --worker-index 0 --worker-count 8
+ *   node -r esbuild-register scripts/backfill-account-authorizations/index.ts --worker-index 1 --worker-count 8
+ *   # ... up to worker-index 7
+ *   # Each worker writes its own checkpoint: temp/backfill-cursor-worker-0.txt, etc.
+ *
+ *   # Filter to a single service
+ *   node -r esbuild-register scripts/backfill-account-authorizations/index.ts --service relay
```
[the remark this answers is not in the document]
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: already fixed this

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 12]
```diff
@@ -0,0 +1,618 @@
...
+    startCursor = cursors.startCursor;
+    endCursor = cursors.endCursor;
+    checkpointFile = `./temp/backfill-cursor-worker-${workerIndex}.txt`;
+    log.info('backfill.worker', { workerIndex, workerCount, checkpointFile });
+  } else {
+    startCursor = program.startCursor
+      ? Buffer.from(program.startCursor, 'hex')
+      : Buffer.alloc(32); // all zeros — before the first token in sort order
+    endCursor = program.endCursor
+      ? Buffer.from(program.endCursor, 'hex')
```
[the remark this answers is not in the document]
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: should be fixed too

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 6 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: Are you saying we should output this message on startup? I'm fine with that, only concern is the amount of logs that are output (one line for each 1k batch) it could get lost. I suppose having it in both places isn't bad

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 8 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:241]: If we make it async we still have to await to ensure it's written before continuing.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 13]
```diff
@@ -0,0 +1,618 @@
...
+    throw new Error(
+      `--worker-index must be 0–${workerCount - 1} when --worker-count is ${workerCount}`
+    );
+  }
+
+  const bytesPerWorker = Math.ceil(256 / workerCount);
+  const firstByteStart = workerIndex * bytesPerWorker;
+  const firstByteEnd = Math.min(firstByteStart + bytesPerWorker - 1, 255);
+
+  // startCursor is an exclusive lower bound (WHERE token > startCursor).
```
[the remark this answers is not in the document]
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:292]: I mean, valid, but also really unlikely we're going to run with more than a handful of workers so I'm not too worried about it

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 10 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:273]: So, I was doing some more looking at the whole 'multi worker' thing, and I'm not sure it'll actually work in practice in stage/prod. It would require spinning up several pods and distributing across them in that way, and the other problem is that the current cursor-checkpoint file that gets written is lost when the pod is spun down.

I had it on my todo to look at it further sooo yeah, let me dig in more to understand how it'll work in practice since we'll have to configure that in the k8 config and get back to you. 

From initial estimates fwiw, if we just work with 1 worker to help prevent overpressure on the DB, it would take ~2 hours to get through a backfill. Adding workers obviously speeds that up, but can put more pressure on the oauth db, specifically the refreshTokens table. Maybe we keep the parallel worker aspect and test behavior on stage to understand performance impact?

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 9 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:235]: Yeahhh, something I noticed when reviewing again too... going to throw it at Claude to see if there's a quick solution!

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 10 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:273]: Circling back. So yeah, the current setup to actually use multiple workers on stage or prod will be a bit more difficult with setup in infra. I think, realistically we should run this on stage with a single worker and see what performance looks like. We can leave the setup for parallel workers in, and if we think running against prod is going to take too long we can test the parallel workers on stage as well

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts [hunk 14]
```diff
@@ -0,0 +1,782 @@
...
+  const placeholders = rows.map(() => '(?, ?, ?, ?)').join(', ');
+  const values = rows.flatMap((r) => [
+    r.uid,
+    r.scope,
+    r.service,
+    r.authorizedAt,
+  ]);
+
+  await query(
+    `INSERT INTO accountAuthorizations (uid, scope, service, authorizedAt)
```
Comment (vbudhram) [inline packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:230]: This was called in my AI review, to use `INSERT IGNORE` so that the row is not locked at all

```
  Production already uses INSERT IGNORE for new writes. Backfill matching production semantics
  means there's only one mental model for authorizedAt: "first observation wins." That's easier to
   reason about than "first observation wins for new rows; LEAST across observations for
  backfilled rows."
```

I'm ok with either since I think the risk is pretty low.

**Review by vbudhram — APPROVED** [review]
@nshirley I pulled this down and ran it with seed test data, and all appears as expected. I think testing this in stage with a single worker will be suffice to verify the data.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 6 above)
Comment (LZoog) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts]: I was thinking so, but yeah if it gets garbled up anyway then eh. Especially if you're the one running this you know about the option and I'm sure if it exits then someone else would read the script details/comments/options.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 10 above)
Comment (LZoog) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:273]: SGTM. 2 hours is not as bad as I thought it would be either.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 14 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:230]: The order of operations on this means we have a chance of finding a token with a more recent authorizedAt and inserting that before the 'first/oldest'. The insert uses an `ON DUPLICATE KEY UPDATE` clause to handle this and try to replicate the same behavior. Since production will always be rolling forward, the first instance we see is likely the 'most recent', but for the sake of consistency and making sure we have accurate data I went this route of the `ON DUPLICATE...`

But, I'm indifferent on this! If it makes sense to just insert whichever we see first and ignore others, I'm good with that and can make that update!

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 9 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:235]: Updated to use a `withRetry` _AND_ a connection pool to handle both transient errors, and network errors. If this does run for long periods it should be more resilient now

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 10 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:273]: Caveat that the 2hour mark is a current estimate, I'll have a better idea once we get this on stage to test!

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts (hunk 7 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations.ts:667]: Yes, when `--worker-count` is present, `checkpointFile` is overridden to `./temp/backfill-cursor-worker-${workerIndex}.txt` regardless of `--checkpoint-file`.

File: packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md (hunk 4 above)
Comment (nshirley) [reply packages/fxa-auth-server/scripts/backfill-account-authorizations/backfill-account-authorizations-test-cases.md:39]: I think what you're describing is covered by case `T07` 'Multi-service' above. One token with multiple scopes produces two authorization records. This case is covering the scenario where a token is found with an odd/mismatching scope we don't know about. It's something we can't assume the intended scope so it's skipped. I doubt we'd have real world cases like this, but worth accounting and testing for

**Comment by nshirley** [comment]
Since the underlying table is going through a more significant overhaul, I'm going to close this and just re-open when ready
